File: Leetcode/dfs_recursion_backtracking/89_GrayCode/Solution.java
class Solution {
    public List<Integer> grayCode(int n) {
        HashSet<Integer> visited = new HashSet<>();
        LinkedList<Integer> res = new LinkedList<>();
        visited.add(0);
        res.add(0);
        dfs(n, res, visited);
        return res;
    }
    private boolean dfs(int n, LinkedList<Integer> res, HashSet<Integer> visited){
        if(visited.size()==(int)Math.pow(2, n))
            return true;
        int last = res.get(res.size()-1);
        for(int i = 0; i<n; ++i){
            int next = (last ^ (1<<i));
            if(!visited.contains(next) && next < (int)Math.pow(2, n)){
                visited.add(next);
                res.add(next);
                if(dfs(n, res, visited))
                    return true;
                res.pollLast();
                visited.remove(next);
            }
        }
        return false;
    }
}

File: Leetcode/dfs_recursion_backtracking/89_GrayCode/Solution.py
class Solution(object):
    def grayCode(self, n):
        """
        :type n: int
        :rtype: List[int]
        """
        visited = set()
        def dfs():
            if len(visited)==2**n:
                return True
            for i in range(n):
                nxt = (res[-1]^(1<<i))
                if nxt in visited or nxt >= 2**n:
                    continue
                visited.add(nxt)
                res.append(nxt)
                if dfs():
                    return True
                res.pop()
                visited.remove(nxt)
        res = [0]
        visited.add(0)
        dfs()
        return res
